mhlib/appSource.go

```diff
@@ -14,7 +14,79 @@
 
 package mhlib
 
+import (
+	"fmt"
+	"path"
+	"path/filepath"
+	"strings"
+)
+
+// AppSourceConfig is what can be defined in a mh configuration file and is used
+// create an AppSource struct. It defines where mh looks for app files.
+//
+// Todo: Define kinds of sources here.
+type AppSourceConfig struct {
+	Kind   string `yaml:"kind"`
+	Name   string `yaml:"name"`
+	Source string `yaml:"source"`
+}
+
+// AppSourceConfigs is an Array of AppSourceConfigs defined in a mh configuration file
+type AppSourceConfigs []AppSourceConfig
+
+// AppSource contains an AppSourceConfig and all AppFiles retrieved via it.
 type AppSource struct {
-	Kind   string
-	Source string
+	AppSourceConfig
+	Files AppFiles
+}
+
+// File takes an AppConfig and sets its File to the matching entry of the
+// AppSource if one exists. Returns false if none exists.
+func (as *AppSource) File(appConfig *AppConfig) bool {
+	if file, ok := as.Files[appConfig.Name]; ok {
+		appConfig.File = &file
+
+		return true
+	}
+
+	return false
+}
+
+// NewAppSource returns an AppSource based on a given AppSourceConfig and path to
+// mh configuration(for "configPath" kind of AppSources).
+//
+// Todo: Get rid of "configFile" argument in favor of defined kinds of
+// AppSources - see AppSourceConfig.
+func NewAppSource(config AppSourceConfig, configFile string) (*AppSource, error) {
+	// Build glob pattern depending on kind of AppSource
+	var pattern string
+	if config.Kind == "path" {
+		pattern = config.Source + "/*.y*ml"
```

Comment:
we use magic strings a lot.  Would prefer something like...
```
package main

import (
	"log"

	yaml "gopkg.in/yaml.v2"
)

type ConfigKind string

var Path ConfigKind = "path"
var ConfigPath ConfigKind = "configPath"

type AppSourceConfig struct {
	Kind   ConfigKind `yaml:"kind"`
	Name   string     `yaml:"name"`
	Source string     `yaml:"source"`
}

// AppSourceConfigs is an Array of AppSourceConfigs defined in a mh configuration file
type AppSourceConfigs []AppSourceConfig

func main() {
	example1 := `
kind: "path"
name: "name1"
source: "source1"
`
	example2 := `
kind: "configPath"
name: "name2"
source: "source2"
`
	var sc1 AppSourceConfig
	var sc2 AppSourceConfig
	yaml.Unmarshal([]byte(example1), &sc1)
	yaml.Unmarshal([]byte(example2), &sc2)
	for _, config := range []AppSourceConfig{sc1, sc2} {
		switch config.Kind {
		case Path:
			log.Printf("%s is path kind.", config.Name)
		case ConfigPath:
			log.Printf("%s is ConfigPath kind.", config.Name)
		}
	}
}
```
output...
```
API server listening at: 127.0.0.1:37022
2018/07/03 09:01:14 name1 is path kind.
2018/07/03 09:01:14 name2 is ConfigPath kind.
```

I see this string typing done a lot in the kubernetes API's.

Reply:
I agree on that we need a change for that. When writing this I was thing of duck-typing and ditching `kind` completely later in favor for pointer attributes like `configPath` and `path`, compare https://godoc.org/k8s.io/api/core/v1#PersistentVolumeSource .